ready for review (#45)
rstudio/tensorflow-blog: _posts/2019-07-30-censored-data/censored_data_tfp.Rmd

author:
  - name: Sigrid Keydana
    affiliation: RStudio
    affiliation_url: https://www.rstudio.com/
slug: keydana2019tfpcensored
date: 07-30-2019
categories:
  - Probability and statistics
output:
  distill::distill_article:
    self_contained: false
preview: images/thumb_cropped.png
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
```

Nothing's ever perfect, and data isn't either. One type of "imperfection" is _missing data_, where some features are unobserved for some subjects. (A topic for another post.) Another is _censored data_, where an event whose characteristics we want to measure does not occur in the observation interval. The example in Richard McElreath's _Statistical Rethinking_ is time to adoption of cats in an animal shelter. If we fix an interval and observe wait times for those cats that actually _did_ get adopted, our estimate will end up too optimistic: We don't take into account those cats who weren't adopted during this interval and thus, would have contributed wait times of length longer than the complete interval.

> In this post, we use a slightly less emotional example which nonetheless may be of interest, especially to R package developers: time to completion of `R CMD check`, collected from CRAN and provided by the `parsnip` package as `check_times`. Here, the censored portion are those checks that errored out for whatever reason, i.e., for whom the check did not complete.

Reviewer: ```suggestion
In this post, we use a slightly less emotional example which nonetheless may be of interest, especially to R package developers: time to completion of `R CMD check`, collected from CRAN and provided by the `parsnip` package as `check_times`. Here, the censored portion are those checks that errored out for whatever reason, i.e., for which the check did not complete.
```

> Why do we care about the censored portion? In the cat adoption scenario, this is pretty obvious: We want to be able to get a realistic estimate for any unknown cat, not just those cats that will turn out to be "lucky". How about `check_times`? Well, if your submission is one of those that errored out, you still care about how long you wait, so even though their percentage is low (< 1%) we don't want to simply exclude them. Also, there is the possibility that the failing ones would have taken longer, had they run to completion, dut to some intrinsic difference between both groups. Conversely, if failures were random, the longer-running checks would have a greater chance to get hit by an error. So here too, exluding the censored data may result in bias.

Reviewer 2: `dut to some ` -> `due to some`

Author: thanks!

> Why do we care about the censored portion, the entities for whom the event of interest did not occur in the period of observation? In the cat adoption scenario, this is pretty obvious: We want to be able to get a realistic estimate for any unknown cat, not just those cats that will turn out to be "lucky". How about `check_times`? Well, if your submission is one of those that errored out, you still care about how long you wait, so even though their percentage is low (< 1%) we don't want to simply exclude them. Also, there is the possibility that the failing ones would have taken longer, had they run to completion, dut to some intrinsic difference between both groups. Conversely, if failures were random, the longer-running checks would have a greater chance to get hit by an error. So here too, exluding the censored data may result in bias.

Reviewer: I would open this with the question: "Why do we care about the censored portion?" Then describe the cat scenario (as you do now), _then_ go into the check times/entities for which event of interest didn't occur.

Author: Hm, I think that is the order I'm following?
I mean, 

"the entities for whom the event of interest did not occur in the period of observation?"

is really nothing than an "extended paraphrase"?

Reviewer: Yeah — it's extended enough that it makes it a bit confusing to have the cats in the middle, and then go back to it.

Author: Yeah you're right, I'm probably overdoing it rephrasing things times & again ... I  think I'll just remove the paraphrase completely, right?

For the others, all we know is that in a virtual world where the check completed, it would take _at least as long_ as the given duration. This quantity can be modeled by the exponential complementary cumulative distribution function (CCDF). Why? A cumulative distribution function (CDF) indicates the probability that a value lower or equal to some reference point was reached; e.g., "the probability of durations <= 255 is 0.9". Its complement, 1 - CDF, then gives the probability that a value will exceed than that reference point.

Let's see this in action.

## The data

The following code works with the current stable releases of TensorFlow and TensorFlow Probability, which are 1.14 and 0.7, respectively. If you don't have `tfprobability`  installed, get it from Github:

```{r}
remotes::install_github("rstudio/tfprobability")
```

These are the libraries we need. As of TensorFlow 1.14, we call `tf$compat$v2$enable_v2_behavior()` to run with eager execution.

```{r}
library(tensorflow)
library(tfprobability)
library(parsnip)
library(tidyverse)
library(zeallot)
library(gridExtra)
library(HDInterval)

tf$compat$v2$enable_v2_behavior()
```

> Besides the check durations we want to model, `check_times` reports various features of the package in question, such as number of imported packages, number of dependencies, size of code and documentation files, etc. The `status` variable indicates whether the check completed or errored out.

Reviewer: ```suggestion
Besides the check durations we want to model, `check_times` reports' various features of the package in question, such as number of imported packages, number of dependencies, size of code and documentation files, etc. The `status` variable indicates whether the check completed or errored out.
```

Author: I think this is a misunderstanding - "reports" is supposed to be read as a verb - had I better choose a different one?

Reviewer: (>ლ) No, that makes total sense — I think I wasn't reading in "code mode"!

Author: okay thanks!

$ ns_import      <dbl> 3, 15, 6, 0, 4, 5, 0, 4, 2, 10, 5, 6, 1, 0, 2, 2, 1, 11, 0,…
$ ns_export      <dbl> 0, 19, 0, 0, 10, 0, 0, 2, 0, 9, 3, 4, 0, 1, 10, 0, 16, 0, 2…
$ s3_methods     <dbl> 3, 0, 11, 0, 0, 0, 0, 2, 0, 23, 0, 0, 2, 5, 0, 4, 0, 0, 0, …
$ s4_methods     <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,…
$ doc_count      <int> 0, 3, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,…
$ doc_size       <dbl> 0.000000, 0.019757, 0.038281, 0.000000, 0.007874, 0.000000,…
$ src_count      <int> 0, 0, 0, 0, 0, 0, 0, 2, 0, 5, 3, 0, 0, 0, 0, 0, 0, 54, 0, 0…
$ src_size       <dbl> 0.000000, 0.000000, 0.000000, 0.000000, 0.000000, 0.000000,…
$ data_count     <int> 2, 0, 0, 3, 3, 1, 10, 0, 4, 2, 2, 146, 0, 0, 0, 0, 0, 10, 0…
$ data_size      <dbl> 0.025292, 0.000000, 0.000000, 4.885864, 4.595504, 0.006500,…
$ testthat_count <int> 0, 8, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 3, 3, 0, 0,…
$ testthat_size  <dbl> 0.000000, 0.002496, 0.000000, 0.000000, 0.000000, 0.000000,…
$ check_time     <dbl> 49, 101, 292, 21, 103, 46, 78, 91, 47, 196, 200, 169, 45, 2…
$ status         <dbl> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,…
```

> Of these 13626 observations, just 103 are censored:

Reviewer: ```suggestion
Of these 13,626 observations, just 103 are censored:
```

```{r}
table(df$status)
```

```
0     1 
103 13523 
```

> For better readability of this text, we want to work with an interesting subset of the columns. We use `lm` to help us guide our selection:

Reviewer: ```suggestion
For better readability, we'll work with an interesting subset of the columns. We use `lm` to help us guide our selection:
```

descr            0.012181   0.002510   4.854 1.23e-06 ***
r_count          0.148266   0.037318   3.973 7.13e-05 ***
r_size         113.029030   5.798490  19.493  < 2e-16 ***
ns_import        0.658115   0.058568  11.237  < 2e-16 ***
ns_export       -0.212985   0.021317  -9.991  < 2e-16 ***
s3_methods      -0.002599   0.028722  -0.090  0.92791    
s4_methods       0.110948   0.161135   0.689  0.49112    
doc_count        8.230161   0.725361  11.346  < 2e-16 ***
doc_size       547.271957  32.644602  16.765  < 2e-16 ***
src_count        1.328554   0.076582  17.348  < 2e-16 ***
src_size        -2.285533   1.583432  -1.443  0.14893    
data_count       0.082173   0.074296   1.106  0.26873    
data_size        3.878125   0.940686   4.123 3.77e-05 ***
testthat_count   0.162082   0.088090   1.840  0.06580 .  
testthat_size    1.080599   0.839307   1.287  0.19795    
status          11.135893   7.609893   1.463  0.14340    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 76.6 on 13602 degrees of freedom
Multiple R-squared:  0.4282,	Adjusted R-squared:  0.4273 
F-statistic: 442.9 on 23 and 13602 DF,  p-value: < 2.2e-16

```

> It seems that if we choose `imports`, `depends`, `r_size`, `doc_size`, `ns_import` and `ns_export` we end up with a mix of (comparatively) powerful predictors from different semantic spaces and of different scales.

Reviewer: ```suggestion
It seems that if we choose `imports`, `depends`, `r_size`, `doc_size`, `ns_import` and `ns_export`, we end up with a mix of (comparatively) powerful predictors from different semantic spaces and of different scales.
```

Before pruning the dataframe, we save away the target variable. In our model and training setup, it is convenient to have censored and uncensored data stored separately, so here we create _two_ target matrices instead of one:

```{r}
# check times for failed checks
# _c stands for censored
check_time_c <- df %>%
  filter(status == 0) %>%
  select(check_time) %>%
  as.matrix()

# check times for successful checks 
check_time_nc <- df %>%
  filter(status == 1) %>%
  select(check_time) %>%
  as.matrix()
```

> Now we can zoom in on the variables of interest, setting up one dataframe for the censored data and one for the uncensored data each. We add a column of `1`s for use as an intercept.

Reviewer: ```suggestion
Now we can zoom in on the variables of interest, setting up one data frame for the censored data and one for the uncensored data. We add a column of `1`s for use as an intercept.
```

| completed                   |     0   |     1   |     3   |     6   |    19  |     312   |
| not completed               |     0   |     2   |     5   |    11   |    23  |     297   |
:::

Same pattern for `r_size`, the size on disk of files in the `R` directory:

::: l-body
| percentiles: _r_size_       |   10%   |   30%   |   50%   |   70%   |   90%  |    100%   |
|-----------------------------|---------|---------|---------|---------|--------|-----------|
| completed                   |   0.005 |   0.015 |   0.031 |   0.063 |  0.176 |     3.746 |
| not completed               |   0.008 |   0.019 |   0.041 |   0.097 |  0.217 |     2.148 |
:::

And finally, we see it for `doc_size` too, where `doc_size` is the size of `.Rmd` and `.Rnw` files:

::: l-body
| percentiles: _doc_size_     |   10%   |   30%   |   50%   |   70%   |   90%  |    100%   |
|-----------------------------|---------|---------|---------|---------|--------|-----------|
| completed                   |   0.000 |   0.000 |   0.000 |   0.000 |  0.023 |     0.988 |
| not completed               |   0.000 |   0.000 |   0.000 |   0.011 |  0.042 |     0.114 |
:::

> Given our task at hand -- model check durations taking into account uncensored as well as censored data -- we won't dwell on differences between both groups any longer; nonetheless we thought it interesting to relate these numbers.

Reviewer: ```suggestion
Given the task at hand -- model check durations taking into account uncensored as well as censored data -- we won't dwell on differences between both groups any longer; nonetheless we thought it interesting to relate these numbers.
```

So now, back to work. We need to create a model.

## The model

> As explained in the introduction, for completed checks duration is modeled using an exponential PDF. This is as straightforward as adding [tfd_exponential()](https://rstudio.github.io/tfprobability/reference/tfd_exponential.html) to the model function, [tfd_joint_distribution_sequential()](https://rstudio.github.io/tfprobability/reference/tfd_joint_distribution_sequential.html). ^[For a first introduction to MCMC sampling with `tfprobability`, see [Tadpoles on TensorFlow: Hierarchical partial pooling with tfprobability](https://blogs.rstudio.com/tensorflow/posts/2019-05-06-tadpoles-on-tensorflow/)]For the censored portion, we need the exponential CCDF. This one is not, as of today, easily added to the model. What we can do though is calculate its value ourselves and add it to the "main" model likelihood. We'll see this below when discussing sampling; for now it means the model definition ends up straightforward as it only covers the non-censored data. It is made of just the said exponential PDF and priors for the regression parameters.

Reviewer: I think this is the first time you've used PDF here, so you might want to spell it out as well.

> As explained in the introduction, for completed checks duration is modeled using an exponential PDF. This is as straightforward as adding [tfd_exponential()](https://rstudio.github.io/tfprobability/reference/tfd_exponential.html) to the model function, [tfd_joint_distribution_sequential()](https://rstudio.github.io/tfprobability/reference/tfd_joint_distribution_sequential.html). ^[For a first introduction to MCMC sampling with `tfprobability`, see [Tadpoles on TensorFlow: Hierarchical partial pooling with tfprobability](https://blogs.rstudio.com/tensorflow/posts/2019-05-06-tadpoles-on-tensorflow/)]For the censored portion, we need the exponential CCDF. This one is not, as of today, easily added to the model. What we can do though is calculate its value ourselves and add it to the "main" model likelihood. We'll see this below when discussing sampling; for now it means the model definition ends up straightforward as it only covers the non-censored data. It is made of just the said exponential PDF and priors for the regression parameters.

Reviewer: ```suggestion
As explained in the introduction, for completed checks duration is modeled using an exponential PDF. This is as straightforward as adding [tfd_exponential()](https://rstudio.github.io/tfprobability/reference/tfd_exponential.html) to the model function, [tfd_joint_distribution_sequential()](https://rstudio.github.io/tfprobability/reference/tfd_joint_distribution_sequential.html). ^[For a first introduction to MCMC sampling with `tfprobability`, see [Tadpoles on TensorFlow: Hierarchical partial pooling with tfprobability](https://blogs.rstudio.com/tensorflow/posts/2019-05-06-tadpoles-on-tensorflow/)]For the censored portion, we need the exponential CCDF. This one is not, as of today, easily added to the model. What we can do though is calculate its value ourselves and add it to the "main" model likelihood. We'll see this below when discussing sampling. For now it means the model definition ends up straightforward as it only covers the non-censored data. It is made of just the exponential PDF and priors for the regression parameters.
```

Reviewer: Removed extraneous "the" from last sentence

> As to the latter, we use 0-centered, wide, Gaussian priors for all parameters. As the priors are all the same, instead of listing a bunch of `tfd_normal`s, we can create them all at once as

Reviewer: ```suggestion
As for the latter, we use 0-centered, wide, Gaussian priors for all parameters. As the priors are all the same, instead of listing a bunch of `tfd_normal`s, we can create them all at once as
```

```{r}
x <- df_nc[1, ]
x
```

```
# A tibble: 1 x 7
  intercept depends imports doc_size r_size ns_import ns_export
      <dbl>   <dbl>   <dbl>    <dbl>  <dbl>     <dbl>     <dbl>
1         1       3       0        0 0.0291         3         0
```

Then for this sample, our model's exponential rate will look like this:

```{r}
1/sum(x * betas)
```

which is `-0.005412808`. Negative rates are not defined for the exponential distribution, and the resulting log probability will be `NaN`.

For the rate to be positive always, all possible linear combinations of the predictors have to end up positive. As the predictors themselves are positive throughout, the parameter samples have to be as well. So if we want to use normal priors, we need to add an extra transformation -- namely, an exponential bijector that ensures MCMC can work in unconstrained space while the model will receive positive samples:

```{r}
unconstraining_bijectors <- list(tfb_exp(), tfb_identity())
```

The next thing to do is define the target we want to optimize.

## Optimization target

Abstractly, the thing to maximize is the log probility of the data -- that is, the measured durations -- under the model.

> Now here the data comes in two parts, and the target does as well. First, we have the non-censored data, for whom

Reviewer: ```suggestion
Now here the data comes in two parts, and the target does as well. First, we have the non-censored data, for which
```

```{r}
m %>% tfd_log_prob(list(betas, tf$cast(target_nc, betas$dtype)))
```

will calculate the log probability. Second, to obtain log probability for the censored data we write a custom function that calculates the log of the exponential CCDF:

```{r}
get_exponential_lccdf <- function(betas, data, target) {
  e <-  tfd_independent(tfd_exponential(rate = 1 / tf$transpose(tf$matmul(
    tf$cast(data, betas$dtype), tf$transpose(betas)
  ))),
  reinterpreted_batch_ndims = 1)
  cum_prob <- e %>% tfd_cdf(tf$cast(target, betas$dtype))
  tf$math$log(1 - cum_prob)
}
```

> Both parts are combined in a little wrapper function that allows to compare training including and excluding the censored data. We won't do that in this post, but you might be interested to do it with your own data, especially if the ratio of censored and uncensored parts is a little less imbalanced.

Reviewer: ```suggestion
Both parts are combined in a little wrapper function that allows us to compare training including and excluding the censored data. We won't do that in this post, but you might be interested to do it with your own data, especially if the ratio of censored and uncensored parts is a little less imbalanced.
```

    num_results = n_steps,
    num_burnin_steps = n_burnin,
    current_state = tf$ones_like(initial_betas),
    trace_fn = trace_fn
  )
}

# important for performance: run HMC in graph mode
run_mcmc <- tf_function(run_mcmc)

res <- hmc %>% run_mcmc()
samples <- res$all_states
```

## Results

Before we inspect the chains, here is a quick look at the proportion of accepted steps and the per-parameter mean step size:

```{r}
accepted <- res$trace[[1]]
as.numeric(accepted) %>% mean()
```

```
0.838
```

```{r}
step_sizes <- res$trace[[2]]
step_sizes %>% as.matrix() %>% apply(2, mean)
```

```
[1] 0.01667838 0.01667838 0.01667838 0.08339182 0.02501755 0.02501755 0.05003509
```

We also store away effective sample sizes and the _rhat_ metrics for later addition to the synopsis. 

```{r}
effective_sample_size <- mcmc_effective_sample_size(samples) %>%
  as.matrix() %>%
  apply(2, mean)
potential_scale_reduction <- mcmc_potential_scale_reduction(samples) %>%
  as.numeric()
```

We then convert the `samples` tensor to an R array for use in postprocessing.

```{r}
# 2-item list, where each item has dim (1000, 4)
samples <- as.array(samples) %>% array_branch(margin = 3)
```

> Now, how well did the sampling work? The chains mix well, but for some parameters, autocorrelation is still pretty high, individual tuning of step sizes nonwithstanding.

Reviewer: ```suggestion
How well did the sampling work? The chains mix well, but for some parameters, autocorrelation is still pretty high, individual tuning of step sizes notwithstanding.
```

```{r}
mse_lm <- ((comp_nc$predicted_lm - comp_nc$actual)^2) %>% mean()
mse_mcmc <- ((comp_nc$predicted_mcmc - comp_nc$actual)^2) %>% mean()
c(mse_lm, mse_mcmc)
```

```
[1] 6273.799 6477.195
```

Doing the same for the censored data:

```{r}
comp_c <- tibble(
  actual = as.vector(check_time_c),
  predicted_lm = as.vector(pred_c_lm),
  predicted_mcmc = as.vector(pred_c)
)
head
```

```
# A tibble: 10 x 3
   actual predicted_lm predicted_mcmc
    <dbl>        <dbl>          <dbl>
 1     10         91.4           91.1
 2     80         87.8           82.3
 3     38         61.0           53.9
 4    304        673.           893. 
 5     89        105.           104. 
 6     68         89.7           92.8
 7      5         71.6           64.3
 8     77         68.0           61.6
 9     89        125.           122. 
10    174        119.           117. 
 
```

```{r}
mse_lm <- ((comp_c$predicted_lm - comp_c$actual) ^ 2) %>% mean()
mse_mcmc <- ((comp_c$predicted_mcmc - comp_c$actual) ^ 2) %>% mean()
c(mse_lm, mse_mcmc)
```

```

[1]  9273.062 11328.088
```

Overall, we conclude that the model works reasonably well.

## Wrapup

> We've shown how to model censored data -- resp. a frequent subtype thereof involving durations -- using `tfprobability`. The `check_times` data from `parsnip` were a fun choice, but this modeling technique may be even more useful when censoring is more substantial. Hopefully his post has provided some guidance on how to handle censored data in your own work. Thanks for reading!

Reviewer: What does `resp.` stand for here?

Author: respectively - I wasn't sure either if this was okay ... I was looking for an alternative to "or" ... would you have an idea?

Author: I could change it to "or rather, a frequent subtype..."?
Would that be better?

Reviewer: Yes!

Author: Okay thanks!
